File: sample-objc/ARGEARSample/UI/UIComponent/Main/Bottom/ModeSelectView.m
//
//  ModeSelectView.m
//  ARGEARSample
//
//  Created by Jihye on 10/09/2019.
//  Copyright © 2019 Seerslab. All rights reserved.
//

#import "ModeSelectView.h"

@implementation ModeSelectView

#define kModeSelectViewFullEnableColor [UIColor whiteColor]
#define kModeSelectViewFullDisableColor [[UIColor whiteColor] colorWithAlphaComponent:0.5f]
#define kModeSelectViewEnablePinkColor [UIColor colorWithRed:48.f/255.f green:99.f/255.f blue:230.f/255.f alpha:1.f]
#define kModeSelectViewDisableGrayColor [UIColor colorWithRed:189.f/255.f green:189.f/255.f blue:189.f/255.f alpha:1.f]

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

- (void)awakeFromNib {
    [super awakeFromNib];
    
    _mode = ARGMediaModePhoto;
    
    NSString *photoString = [@"photo" localized];
    NSString *videoString = [@"video" localized];
    
    [_pictureButton setTitle:photoString forState:UIControlStateNormal];
    [_videoButton setTitle:videoString forState:UIControlStateNormal];
    
    CGSize photoStringSize = [photoString sizeWithAttributes:@{NSFontAttributeName:[UIFont fontWithName:@"NotoSansKR-Regular" size:13.f]}];
    CGSize videoStringSize = [videoString sizeWithAttributes:@{NSFontAttributeName:[UIFont fontWithName:@"NotoSansKR-Regular" size:13.f]}];

    CGFloat viewWidth = 1.5f * photoStringSize.width + 1.5f * videoStringSize.width + 18 * 2 + 5 * 2 + 4;
    [_viewWidth setConstant:viewWidth];
    
    [_pictureButtonWidth setConstant:photoStringSize.width + 2];
    [_videoButtonWidth setConstant:videoStringSize.width + 2];
    
    CGFloat modeButtonsViewWidth = [_pictureButtonWidth constant] + 18.f + [_videoButtonWidth constant];
    [_modeButtonsViewWidth setConstant:modeButtonsViewWidth];
    
    [_underlineViewWidth setConstant:[_pictureButtonWidth constant] + 10];
}

- (IBAction)pictureButtonAction:(UIButton *)sender {
    if (_mode == ARGMediaModeVideo) {
        [self changePhotoMode];
    }
}

- (IBAction)videoButtonAction:(UIButton *)sender {
    if (_mode == ARGMediaModePhoto) {
        [self changeVideoMode];
    }
}

- (IBAction)handleSwipeFrom:(UISwipeGestureRecognizer*)recognizer {
    
    if ([recognizer direction] == UISwipeGestureRecognizerDirectionLeft) {
        if (_mode == ARGMediaModePhoto) {
            [self changeVideoMode];
        }
    } else if ([recognizer direction] == UISwipeGestureRecognizerDirectionRight) {
        if (_mode == ARGMediaModeVideo) {
            [self changePhotoMode];
        }
    }
}

// change to photo mode
- (void)changePhotoMode {
    [self setMode:ARGMediaModePhoto];
    
    __weak ModeSelectView *weakSelf = self;
    CGFloat trailing = [[[[_pictureButton titleLabel] text] lowercaseString] isEqualToString:@"photo"] ? 4 : 2;

    [UIView animateWithDuration:0.2 delay:0.0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
        [[weakSelf modeButtonsViewTrailing] setConstant:trailing];
        [[weakSelf underlineViewWidth] setConstant:[[weakSelf pictureButtonWidth] constant] + 10];

        [weakSelf changeModeColor];
        
        [self layoutIfNeeded];
    } completion:nil];
}

// change to video mode
- (void)changeVideoMode {
    [self setMode:ARGMediaModeVideo];
    
    __weak ModeSelectView *weakSelf = self;
    CGFloat trailing = [[[[_pictureButton titleLabel] text] lowercaseString] isEqualToString:@"photo"] ? 2 : 0;

    [UIView animateWithDuration:0.2 delay:0.0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
        [[weakSelf modeButtonsViewTrailing] setConstant:2 + [[weakSelf pictureButtonWidth] constant]/2 + 18 + [[weakSelf videoButtonWidth] constant]/2 + trailing];
        [[weakSelf underlineViewWidth] setConstant:[[weakSelf videoButtonWidth] constant] + 10];
        
        [weakSelf changeModeColor];
        
        [self layoutIfNeeded];
    } completion:nil];
}

- (void)changeModeColor {
    if ([[_underlineView backgroundColor] isEqual:kModeSelectViewFullEnableColor]) {
        // full
        if (_mode == ARGMediaModePhoto) {
            
            [_pictureButton setTitleColor:kModeSelectViewFullEnableColor forState:UIControlStateNormal];
            [_videoButton setTitleColor:kModeSelectViewFullDisableColor forState:UIControlStateNormal];
        } else {

            [_pictureButton setTitleColor:kModeSelectViewFullDisableColor forState:UIControlStateNormal];
            [_videoButton setTitleColor:kModeSelectViewFullEnableColor forState:UIControlStateNormal];
        }
    } else {
        // 4:3, 1:1
        if (_mode == ARGMediaModePhoto) {
            
            [_pictureButton setTitleColor:kModeSelectViewEnablePinkColor forState:UIControlStateNormal];
            [_videoButton setTitleColor:kModeSelectViewDisableGrayColor forState:UIControlStateNormal];
        } else {
            
            [_pictureButton setTitleColor:kModeSelectViewDisableGrayColor forState:UIControlStateNormal];
            [_videoButton setTitleColor:kModeSelectViewEnablePinkColor forState:UIControlStateNormal];
        }
    }
}

- (void)setRatio:(ARGMediaRatio)ratio {
    
    if (ratio == ARGMediaRatio_16x9) {
        
        [_underlineView setBackgroundColor:kModeSelectViewFullEnableColor];
    } else {

        [_underlineView setBackgroundColor:kModeSelectViewEnablePinkColor];
    }
    [self changeModeColor];
}

@end

File: sample-swift/ARGearSample/UI/UIComponent/Main/Bottom/Camera/ModeSelectView.swift
//
//  ModeSelectView.swift
//  ARGearSample
//
//  Created by Jihye on 2020/01/30.
//  Copyright © 2020 Seerslab. All rights reserved.
//

import UIKit
import ARGear

class ModeSelectView: UIView {
    
    let kModeSelectViewFullEnableColor = UIColor.white
    let kModeSelectViewFullDisableColor = UIColor.white.withAlphaComponent(0.5)
    let kModeSelectViewEnablePinkColor = UIColor(red: 48.0/255.0, green: 99.0/255.0, blue: 230.0/255.0, alpha: 1.0)
    let kModeSelectViewDisableGrayColor = UIColor(red: 189.0/255.0, green: 189.0/255.0, blue: 189.0/255.0, alpha: 1.0)
    
    @objc dynamic var mode: ARGMediaMode = .photo
    
    @IBOutlet weak var viewWidth: NSLayoutConstraint!
    @IBOutlet weak var pictureButtonWidth: NSLayoutConstraint!
    @IBOutlet weak var videoButtonWidth: NSLayoutConstraint!
    @IBOutlet weak var modeButtonsViewWidth: NSLayoutConstraint!

    @IBOutlet weak var modeButtonsView: UIView!
    @IBOutlet weak var modeButtonsViewTrailing: NSLayoutConstraint!

    @IBOutlet weak var pictureButton: UIButton!
    @IBOutlet weak var videoButton: UIButton!

    @IBOutlet weak var underlineView: UIView!
    @IBOutlet weak var underlineViewWidth: NSLayoutConstraint!
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        let photoString = "photo".localized()
        let videoString = "video".localized()
        
        self.pictureButton.setTitle(photoString, for: .normal)
        self.videoButton.setTitle(videoString, for: .normal)
        
        let photoStringSize = photoString.size(withAttributes: [.font: UIFont(name: "NotoSansKR-Regular", size: 13.0)!])
        let videoStringSize = videoString.size(withAttributes: [.font: UIFont(name: "NotoSansKR-Regular", size: 13.0)!])
        
        let viewWidth = 1.5 * photoStringSize.width + 1.5 * videoStringSize.width + 18 * 2 + 5 * 2 + 4
        self.viewWidth.constant = viewWidth
        
        self.pictureButtonWidth.constant = photoStringSize.width + 2
        self.videoButtonWidth.constant = videoStringSize.width + 2

        let modeButtonsViewWidth = self.pictureButtonWidth.constant + 18.0 + self.videoButtonWidth.constant
        self.modeButtonsViewWidth.constant = modeButtonsViewWidth
        
        self.underlineViewWidth.constant = self.pictureButtonWidth.constant + 10
    }
    
    func setRatio(_ ratio: ARGMediaRatio) {
        
        var underlineViewColor = kModeSelectViewEnablePinkColor
        if ratio == ._16x9 {
            underlineViewColor = kModeSelectViewFullEnableColor
        }
        self.underlineView.backgroundColor = underlineViewColor
        
        self.changeModeColor()
    }
    
    @IBAction func pictureButtonAction(_ sender: UIButton) {
        if self.mode == .video {
            self.changePhotoMode()
        }
    }
    
    @IBAction func videoButtonAction(_ sender: UIButton) {
        if self.mode == .photo {
            self.changeVideoMode()
        }
    }
    
    @IBAction func handleSwipeFrom(_ swipe: UISwipeGestureRecognizer) {
        if swipe.direction == .left {
            if self.mode == .photo {
                self.changeVideoMode()
            }
        } else if swipe.direction == .right {
            if self.mode == .video {
                self.changePhotoMode()
            }
        }
    }
    
    // change to photo mode
    func changePhotoMode() {
        self.mode = .photo
        
        let trailing = (self.pictureButton.title(for: .normal)?.lowercased() == "photo") ? 4 : 2
        UIView.animate(withDuration: 0.2, delay: 0, options: .curveEaseInOut, animations: {
            self.modeButtonsViewTrailing.constant = CGFloat(trailing)
            self.underlineViewWidth.constant = self.pictureButtonWidth.constant + 10
            
            self.changeModeColor()
            
            self.layoutIfNeeded()
        }, completion: nil)
    }
    
    // change to video mode
    func changeVideoMode() {
        self.mode = .video
        
        let halfPhotoWidth = self.pictureButtonWidth.constant/2
        let halfVideoWidth = self.videoButtonWidth.constant/2
        
        let trailing = (self.pictureButton.title(for: .normal)?.lowercased() == "photo") ? 2 : 0
        UIView.animate(withDuration: 0.2, delay: 0, options: .curveEaseInOut, animations: {
            self.modeButtonsViewTrailing.constant = 2 + halfPhotoWidth + 18 + halfVideoWidth + CGFloat(trailing)
            self.underlineViewWidth.constant = self.videoButtonWidth.constant + 10
            
            self.changeModeColor()
            
            self.layoutIfNeeded()
        }, completion: nil)
    }
    
    func changeModeColor() {

        var picureButtonTitleColor: UIColor = .clear
        var videoButtonTitleColor: UIColor = .clear

        if self.underlineView.backgroundColor == kModeSelectViewFullEnableColor {
            // full
            if self.mode == .photo {
                picureButtonTitleColor = kModeSelectViewFullEnableColor
                videoButtonTitleColor = kModeSelectViewFullDisableColor
            } else {
                picureButtonTitleColor = kModeSelectViewFullDisableColor
                videoButtonTitleColor = kModeSelectViewFullEnableColor
            }
        } else {
            // 4:3, 1:1
            if self.mode == .photo {
                picureButtonTitleColor = kModeSelectViewEnablePinkColor
                videoButtonTitleColor = kModeSelectViewDisableGrayColor
            } else {
                picureButtonTitleColor = kModeSelectViewDisableGrayColor
                videoButtonTitleColor = kModeSelectViewEnablePinkColor
            }
        }
        
        self.pictureButton.setTitleColor(picureButtonTitleColor, for: .normal)
        self.videoButton.setTitleColor(videoButtonTitleColor, for: .normal)
    }
}
